Create initial Entities data model specification. (#4442)
open-telemetry/opentelemetry-specification: specification/resource/data-model.md

# Resource Data Model

**Status**: [Development](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Identity](#identity)
  * [Navigation](#navigation)
  * [Telescoping](#telescoping)

<!-- tocstop -->

</details>

> A Resource is an immutable representation of the entity producing telemetry as Attributes. For example, a process producing telemetry that is running in a container on Kubernetes has a Pod name, it is in a namespace and possibly is part of a Deployment which also has a name. All three of these attributes can be included in the Resource. Note that there are certain "standard attributes" that have prescribed meanings.

Reviewer: > a process producing telemetry that is running in a container on Kubernetes has a Pod name

Does it? This wording might need to be more precise, like, "detects a Pod name."

Author: Ok, I took a crack at this, but still not happy with the clarity of thought.  Advice welcome on wording.

> A resource is composed of [`Entity`](../entities/README.md) and raw attributes.

Reviewer 2: 0 or more entities. I would also define what you mean by "raw attributes"
```suggestion
A resource is composed of 0 or more [`Entities`](../entities/README.md) and 0 or more attributes not associated with any entity.
```

> A Resource is an immutable representation of the entity producing telemetry as Attributes. For example, You could have a process producing telemetry that is running in a container on Kubernetes, which is associated to a Pod running on a Node that is a VM but also is in a namespace and possibly is part of a Deployment. Resource could have attributes to denote information about the Container, the Pod, the Node, the VM or the Deployment. All of these help identify what produced the telemetry. Note that there are certain "standard attributes" that have prescribed meanings.

Reviewer 3: What is the relationship of the Resource to the non-identifying attributes of the Entities it is composed of?
Can/must they be included or must the non-identifying not be included?
A priori including them seems to contradict the requirement for a Resource to be immutable since the non-identifying attributes may change.

Reviewer 2: Resource immutability is considered to be a problem that the entities sig/signal is solving. In the real world resources do change all the time and OTLP receivers must be robust to that fact.

IMO the immutability of resources has always been ambiguous. There are 2 possible interpretations:

1. Code-level immutability. In this interpretation the resource is an immutable object, but the same resource need not necessarily be emitted with every batch of telemetry data.
2. Conceptual immutability. In this interpretation the resource and all its component attributes are a stable identifier for... something (presumably the service, SDK, or process described by the telemetry data).

It's my understanding that the original intention of resource immutability was (2), but in practice it is often not the case. Resources change when processes are restarted and in collector pipelines. Even custom processors sometimes modify resource and violate immutability constraints.

---

In any case, if my understanding is correct we should probably remove the statement "A Resource is an immutable representation of the entity producing telemetry" from this PR.

Reviewer 3: Thanks for explaining this.

I think this should be clarified in the spec here.
Should the `immutable` in `specification/resource/sdk.md` be adjusted as well?

Author: We're in a bit of a pickle here.

`Resource SDK` is stable portions of our specification.  It actually *forces* both 1 and 2.  In practice, we see users leveraging resource in ways that aren't technically compliant with our Specification.  As @dyladan appropriately calls out, we're trying to fix this.

However - The *way* to do this with stable specification is. interesting.  I don't want to modify stable pieces of the specification to rely on unstable portions yet.   

This PR is scoped to simple doing the following:

- Defining the Resource Model and its importance as seen today.  This is *not* a stable specification, as it's actually (somehow) brand new.
- Add in the new Entities data model and its relation to resource.

There will be follow on PRs to address your concerns as we find/create non-stable portions of the specification to match the current design/prototypes for Entiies work and the cleanup we're doing here.

Resource provides two important aspects for observability:

- It MUST *identify* an entity that is producing telemetry.
- It SHOULD allow users to determine *where* that entity resides within their infrastructure.

## Identity

Most resources are a composition of `Entity`. `Entity` is described [here](../entities/data-model.md), and includes its own notion of identity. The identity of a resource is the set
of entities contained within it. Two resources are considered different if one
contains an entity not found in the other.

Some resources include raw attributes in additon to Entities. Raw attributes are
considered identifying on a resource. That is, if the key-value pairs of
raw attributes are different, then you can assume the resource is different.

### Navigation

> Implicit in the design of Resource and attributes is ensuring users are able to navigate their infrastructure, tools, Uis, etc. to find the *same* entity that telemetry is reporting against.  For example, in the definition above, we see a few components listed for one entity:

Reviewer 2: ```suggestion
Implicit in the design of Resource and attributes is ensuring users are able to navigate their infrastructure, tools, UIs, etc. to find the *same* entity that telemetry is reporting against.  For example, in the definition above, we see a few components listed for one entity:
```

> Implicit in the design of Resource and attributes is ensuring users are able to navigate their infrastructure, tools, UIs, etc. to find the *same* entity that telemetry is reporting against.  For example, in the definition above, we see a few components listed for one entity:

Reviewer 4: ```suggestion
Implicit in the design of Resource and attributes is ensuring users are able to navigate their infrastructure, tools, UIs, etc. to find the *same* entity that telemetry is reporting against.  For example, in the definition above, we see a few entities listed for one Resource:
```

- A process
- A container
- A kubernetes pod name
- A namespace
- A deployment

> By including identifying attributes of each of these, we can help users navigate through their `kubectl` or kubernetes UIs to find the specific process generating telemetry.   This is as important as being able to uniquely identify one process from another.

Reviewer: ```suggestion
By including identifying attributes of each of these, we can help users navigate through their `kubectl` or Kubernetes UIs to find the specific process generating telemetry.   This is as important as being able to uniquely identify one process from another.
```

> > Aside: Observability signals SHOULD be actionable.  Knowing a process is struggling is not as useful as > being able to scale up a deployment to take load off the struggling process.

Reviewer: ```suggestion
> Aside: Observability signals SHOULD be actionable.  Knowing a process is struggling is not as useful as being able to scale up a deployment to take load off the struggling process.
```

> > Aside: Observability signals SHOULD be actionable.  Knowing a process is struggling is not as useful as being able to scale up a deployment to take load off the struggling process.

Reviewer 4: This seems a bit irrelevant, not sure why it needs to be in the data model doc.

Author: We have a lack of motivating reasons for why we have Resource in OpenTelemetry.  There's not a good place to hang this otherwise, and it's often repeated content when introducing someone to open telemetry.

For those of us who have been living/breathing OpenTelemetry for a while, I'd agree it's taken for granted.

However it's *never said* or motivated. I think it's absolutely critical we include information like this.  Happy for advise on where to put it.

Reviewer 4: Sure, I think it is could to have this sort of info in the spec, but probably a better place can be something like the README. See for example [logs README](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/README.md) which has a lot of non-normative text similar to this.

Author: Thank @tigrannajaryan - This was a great suggestion.  Moved this to README and I think it flows better now / focuses the markdown in the data model on its core content.

PTAL.

> If the only thing important to Resource was identity, we could simply use UUIDs.

Reviewer 4: I don't think this is entirely true. UUIDs are also bad for understanding/learning even if you don't need to act. The previous paragraph seems to make a claim and tries to justify it, but I think it creates confusion. I would delete the previous and this paragraphs.

Author: I fixed this sentence up into a full a paragraph with transition.  PTAL
